chasm/lib/activity/handler.go

```diff
@@ -53,3 +56,115 @@ func (h *handler) StartActivityExecution(ctx context.Context, req *activitypb.St
+	defer func() {
+		var notFound *serviceerror.NotFound
+		if errors.As(err, &notFound) {
+			// TODO(dan): include execution key in error message; we may do this at the CHASM
+			// framework level.
+			// cf. "workflow execution not found for workflow ID XXX and run ID YYY"
+			err = serviceerror.NewNotFound("activity execution not found")
+		}
+	}()
+
+	waitPolicy := req.GetFrontendRequest().GetWaitPolicy()
+
+	if waitPolicy == nil {
+		return chasm.ReadComponent(ctx, ref, (*Activity).buildPollActivityExecutionResponse, req, nil)
+	}
+
+	// Do not allow long-poll to use all remaining time
+	childCtx := ctx
+	if deadline, ok := ctx.Deadline(); ok {
+		var cancel context.CancelFunc
+		childCtx, cancel = context.WithDeadline(ctx, deadline.Add(-time.Second))
+		defer cancel()
+	}
+
+	switch waitPolicy.(type) {
+	case *workflowservice.PollActivityExecutionRequest_WaitAnyStateChange:
+		token := req.GetFrontendRequest().
+			GetWaitPolicy().(*workflowservice.PollActivityExecutionRequest_WaitAnyStateChange).
+			WaitAnyStateChange.GetLongPollToken()
+		if len(token) == 0 {
+			return chasm.ReadComponent(ctx, ref, (*Activity).buildPollActivityExecutionResponse, req, nil)
+		}
+		response, _, err = chasm.PollComponent(childCtx, ref, func(
+			a *Activity,
+			ctx chasm.Context,
+			req *activitypb.PollActivityExecutionRequest,
+		) (*activitypb.PollActivityExecutionResponse, bool, error) {
+			changed, err := chasm.ExecutionStateChanged(a, ctx, token)
+			if err != nil {
+				if errors.Is(err, chasm.ErrMalformedComponentRef) {
+					return nil, false, serviceerror.NewInvalidArgument("invalid long poll token")
+				}
+				if errors.Is(err, chasm.ErrInvalidComponentRef) {
+					return nil, false, serviceerror.NewInvalidArgument("long poll token does not match execution")
+				}
+				return nil, false, err
+			}
+			if changed {
+				response, err := a.buildPollActivityExecutionResponse(ctx, req)
+				return response, true, err
+			} else {
+				return nil, false, nil
+			}
+		}, req)
+	case *workflowservice.PollActivityExecutionRequest_WaitCompletion:
+		// TODO(dan): add functional test when RecordActivityTaskCompleted is implemented
+		response, _, err = chasm.PollComponent(childCtx, ref, func(
+			a *Activity,
+			ctx chasm.Context,
+			req *activitypb.PollActivityExecutionRequest,
+		) (*activitypb.PollActivityExecutionResponse, bool, error) {
+			// TODO(dan): check for terminal activity states
+			panic("pollActivityExecutionWaitCompletion is not implemented")
+			completed := false
+			if completed {
+				response, err := a.buildPollActivityExecutionResponse(ctx, req)
+				if err != nil {
+					return nil, true, err
+				}
+				return response, true, nil
+			}
+			return nil, false, nil
+		}, req)
+	default:
+		return nil, serviceerror.NewInvalidArgumentf("unexpected wait policy type: %T", waitPolicy)
+	}
+
+	if childCtx.Err() != nil {
+		// Caller deadline exceeded
+		return nil, childCtx.Err()
+	}
+	if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
```suggestion
```

Reply:
Always translate context deadline errors to empty responses.
Maybe with the exception of canceled contexts, need to think about that.

Reply:
If we're not on the same page here then let's first agree on requirements via the test `Test_PollActivityExecution_DeadlineExceeded`. Then once the requirements are agreed we can settle the implementation.